### DISCUSSION TITLE
[Question]: Any plans to update object detection to YOLOv9?

### DISCUSSION THREAD

**Discussion by Lensters** [blakeblackshear/frigate#10624] [Ask A Question]
### What is your question:

Any plans to update object detection to YOLOv9?

**Comment by harakas** [comment]
At this moment it's too new. Also I can't find a pre-trained small or nano sized model for it. From one article I can see that it can have much worse inference speeds compared to yolov8: https://learnopencv.com/yolov9-advancing-the-yolo-legacy/#aioseo-yolov9-vs-yolov8 -- haven't tried myself though.

Let's hold off for a while to observe if it truly delivers on its promises as it matures.

**Reply by harakas** [reply]
For reference, the authors of the model say they will release the smaller models once their article gets accepted and published: https://github.com/WongKinYiu/yolov9/issues/3 -- so lets wait for that.

But I'm not quite sure what to think of such a thing. Then again yolov8 has not been published at all, not even a preprint (https://github.com/ultralytics/ultralytics/issues/1564).

After reading more I believe the slow inference with yolov9 mentioned before is likely due to the "auxiliary training branch" in the model being kept in -- it could be removed to speed the inference up.

**Reply by Lensters** [reply]
It is a little suspicious that they aren't releasing their models.  That isn't normal for academics.

**Comment by blakeblackshear** [comment]
I am not sure if this applies to the yolov9 repository you lined, but see here: https://github.com/blakeblackshear/frigate/pull/10717

@harakas You should also remove your public fork with yolo8 support as you are likely in violation as well.

**Reply by harakas** [reply]
I don't think so. As frigate is under MIT I can simply relicense my fork under AGPL.

```
The MIT License

Copyright (c) 2020 Blake Blackshear

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, **sublicense**, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.
```

**Reply by harakas** [reply]
I have now done so: https://github.com/harakas/frigate?tab=License-1-ov-file

**Reply by blakeblackshear** [reply]
The reason I haven't taken that approach is because it isn't clear to me the degree to which you must now honor their interpretation that all outputs of your fork are now AGPL licensed too. I'm not going to relicense Frigate without serious consideration.

**Reply by harakas** [reply]
I believe it's a good thing that Frigate is under MIT. I will not say more.

**Reply by Lensters** [reply]
Without a copy of v9 to look at and no mention of its lineage in their press releases, it is impossible to say if v9 forked from v8.   It is very possible that v9 is from v7 because they're both from the same authors.

**Reply by harakas** [reply]
> Without a copy of v9 to look at and no mention of its lineage in their press releases, it is impossible to say if v9 forked from v8. It is very possible that v9 is from v7 because they're both from the same authors.

Larger models and training code for yolov9 are available on github (https://github.com/WongKinYiu/yolov9/). As is a preprint of their paper on [arxiv.org](https://arxiv.org/pdf/2402.13616.pdf). On a cursory glance I believe they build on their previous work with yolov7 (AF). So lets not delve into groundless speculation and hyperboles.

After some googling the current consensus seems to be that neural network model structure (node/layer structure) is not copyrightable. I'm sure you'll find a copyright lawyer who argues otherwise -- it's their job after all to try to copyright everything on this earth -- speaking from experience interacting with lawyers ;) But that's just wishful thinking on their part. It might be patentable (within a context of application, as far as I can tell). Also training your model with GPL or AGPL software does not make the resulting model/weights (A)GPL. It's a different thing if you use a pre-existing trained model (as we do with current yolov8 distribution in frigate -- pretrained weights from ultralytics AGPL release).

So as far as I can see one just needs to train brand new weights for yolov8 model from nothing, export an onnx file and that can then be licensed and distributed and used by the person who did it however they wish.

**Comment by felipecrs** [comment]
Can someone share your YOLO v9 ONNX model built with 320x320? I'm having a really hard time trying to build it myself.

**Reply by NickM-27** [reply]
would be curious to know what step doesn't work for you https://deploy-preview-16390--frigate-docs.netlify.app/configuration/object_detectors#yolov9

**Reply by felipecrs** [reply]
When running Ubuntu 24.04 in my WSL2, it fails when running the export command with:

```console
❯ bin/python3 export.py --weights ./yolov9-t.pt --imgsz 320 --simplify --include onnx
export: data=data/coco.yaml, weights=['./yolov9-t.pt'], imgsz=[320], batch_size=1, device=cpu, half=False, inplace=False, keras=False, optimize=False, int8=False, dynamic=False, simplify=True, opset=12, verbose=False, workspace=4, nms=False, agnostic_nms=False, topk_per_class=100, topk_all=100, iou_thres=0.45, conf_thres=0.25, include=['onnx']
YOLO 🚀 v0.1-104-g5b1ea9a Python-3.12.3 torch-2.7.0+cu126 CPU

Fusing layers...
gelan-t summary: 489 layers, 2001840 parameters, 22704 gradients, 7.7 GFLOPs

PyTorch: starting from yolov9-t.pt with output shape (1, 84, 2100) (4.4 MB)
free(): double free detected in tcache 2
[1]    1745 IOT instruction (core dumped)  bin/python3 export.py --weights ./yolov9-t.pt --imgsz 320 --simplify --includ
```

In my Ubuntu 25.04 (native) I could not even get past the pip install phase. It fails during the build of some dependency running some cmake command (cmake is installed).

**Reply by Lensters** [reply]
It fails for me during the onnx install.

`lensters@lightning:~/src/yolov9> bin/pip install onnx onnxruntime onnx-simplifier>=0.4.1
  WARNING: Generating metadata for package onnx-simplifier produced metadata for project name onnxsim. Fix your #egg=onnx-simplifier fragments.
  WARNING: Generating metadata for package onnx-simplifier produced metadata for project name onnxsim. Fix your #egg=onnx-simplifier fragments.
  WARNING: Generating metadata for package onnx-simplifier produced metadata for project name onnxsim. Fix your #egg=onnx-simplifier fragments.
  WARNING: Generating metadata for package onnx-simplifier produced metadata for project name onnxsim. Fix your #egg=onnx-simplifier fragments.
  WARNING: Generating metadata for package onnx-simplifier produced metadata for project name onnxsim. Fix your #egg=onnx-simplifier fragments.
  WARNING: Generating metadata for package onnx-simplifier produced metadata for project name onnxsim. Fix your #egg=onnx-simplifier fragments.
  WARNING: Generating metadata for package onnx-simplifier produced metadata for project name onnxsim. Fix your #egg=onnx-simplifier fragments.
  WARNING: Generating metadata for package onnx-simplifier produced metadata for project name onnxsim. Fix your #egg=onnx-simplifier fragments.
  WARNING: Generating metadata for package onnx-simplifier produced metadata for project name onnxsim. Fix your #egg=onnx-simplifier fragments.
  WARNING: Generating metadata for package onnx-simplifier produced metadata for project name onnxsim. Fix your #egg=onnx-simplifier fragments.
  WARNING: Generating metadata for package onnx-simplifier produced metadata for project name onnxsim. Fix your #egg=onnx-simplifier fragments.
  WARNING: Generating metadata for package onnx-simplifier produced metadata for project name onnxsim. Fix your #egg=onnx-simplifier fragments.
  WARNING: Generating metadata for package onnx-simplifier produced metadata for project name onnxsim. Fix your #egg=onnx-simplifier fragments.
  WARNING: Generating metadata for package onnx-simplifier produced metadata for project name onnxsim. Fix your #egg=onnx-simplifier fragments.
  WARNING: Generating metadata for package onnx-simplifier produced metadata for project name onnxsim. Fix your #egg=onnx-simplifier fragments.
  WARNING: Generating metadata for package onnx-simplifier produced metadata for project name onnxsim. Fix your #egg=onnx-simplifier fragments.
  WARNING: Generating metadata for package onnx-simplifier produced metadata for project name onnxsim. Fix your #egg=onnx-simplifier fragments.
  WARNING: Generating metadata for package onnx-simplifier produced metadata for project name onnxsim. Fix your #egg=onnx-simplifier fragments.
  WARNING: Generating metadata for package onnx-simplifier produced metadata for project name onnxsim. Fix your #egg=onnx-simplifier fragments.
  error: subprocess-exited-with-error
  
  × Preparing metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [34 lines of output]
      <string>:26: DeprecationWarning: Use shutil.which instead of find_executable
      fatal: not a git repository (or any of the parent directories): .git
      fatal: not a git repository (or any of the parent directories): .git
      Traceback (most recent call last):
        File "/home/lharold/src/yolov9/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
          main()
        File "/home/lharold/src/yolov9/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/home/lharold/src/yolov9/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 149, in prepare_metadata_for_build_wheel
          return hook(metadata_directory, config_settings)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/tmp/pip-build-env-261vini1/overlay/lib/python3.12/site-packages/setuptools/build_meta.py", line 374, in prepare_metadata_for_build_wheel
          self.run_setup()
        File "/tmp/pip-build-env-261vini1/overlay/lib/python3.12/site-packages/setuptools/build_meta.py", line 512, in run_setup
          super().run_setup(setup_script=setup_script)
        File "/tmp/pip-build-env-261vini1/overlay/lib/python3.12/site-packages/setuptools/build_meta.py", line 317, in run_setup
          exec(code, locals())
        File "<string>", line 271, in <module>
        File "/tmp/pip-build-env-261vini1/overlay/lib/python3.12/site-packages/setuptools/__init__.py", line 117, in setup
          return distutils.core.setup(**attrs)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/tmp/pip-build-env-261vini1/overlay/lib/python3.12/site-packages/setuptools/_distutils/core.py", line 148, in setup
          _setup_distribution = dist = klass(attrs)
                                       ^^^^^^^^^^^^
        File "/tmp/pip-build-env-261vini1/overlay/lib/python3.12/site-packages/setuptools/dist.py", line 334, in __init__
          self.metadata.version = self._normalize_version(self.metadata.version)
                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/tmp/pip-build-env-261vini1/overlay/lib/python3.12/site-packages/setuptools/dist.py", line 370, in _normalize_version
          normalized = str(Version(version))
                           ^^^^^^^^^^^^^^^^
        File "/tmp/pip-build-env-261vini1/overlay/lib/python3.12/site-packages/setuptools/_vendor/packaging/version.py", line 202, in __init__
          raise InvalidVersion(f"Invalid version: {version!r}")
      packaging.version.InvalidVersion: Invalid version: 'unknown'
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
`

**Reply by felipecrs** [reply]
This works for me:

- https://github.com/blakeblackshear/frigate/pull/19224

**Reply by felipecrs** [reply]
Here's the converted model:

[yolov9-t.onnx.txt](https://github.com/user-attachments/files/21333264/yolov9-t.onnx.txt)

Remove the `.txt` extension from it after downloading.

I'm using with:

```yaml
model:
  model_type: yolo-generic
  width: 320
  height: 320
  input_tensor: nchw
  input_dtype: float
  path: /config/yolov9-t.onnx
  labelmap_path: /labelmap/coco-80.txt
```
